### PR TITLE
[WIP] Fixed #31259 -- Added a dark theme in the admin module

### PR DESCRIPTION
**Pull request by mimi89999**
Some parts of the UI aren't themed yet, but most are. Please review code style and colors. If everything is OK, I will theme the entire admin UI.

### CODE REVIEW

File: django/contrib/admin/templates/admin/base.html [hunk 1]
```diff
@@ -4,6 +4,7 @@
 <head>
 <title>{% block title %}{% endblock %}</title>
 <link rel="stylesheet" type="text/css" href="{% block stylesheet %}{% static "admin/css/base.css" %}{% endblock %}">
+<link rel="stylesheet" type="text/css" href="{% block dark-scheme %}{% static "admin/css/dark.css" %}{% endblock %}" media="(prefers-color-scheme: dark)">
```
Comment (carltongibson) [inline django/contrib/admin/templates/admin/base.html]: Nice media query on the link element.

Comment (felixxm) [reply django/contrib/admin/templates/admin/base.html]: Yes, unfortunately it's not supported by [IE or Edge](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) :disappointed:

Comment (carltongibson) [reply django/contrib/admin/templates/admin/base.html]: It's OK. No-one uses those. 😀



('Twas a jest.)

Comment (carltongibson) [reply django/contrib/admin/templates/admin/base.html]: More seriously, support seems to be coming in Edge, and the whole point of these kind of queries is that they degrade harmlessly. A browser that doesn't understand `prefers-color-scheme` will just ignore it.

Comment (felixxm) [reply django/contrib/admin/templates/admin/base.html]: But it means that all IE and Edge browsers will get a dark scheme. Am I right? :thinking:

Comment (carltongibson) [reply django/contrib/admin/templates/admin/base.html]: No, they won't download it.

**Comment by mimi89999** [comment]
I made some screenshots to have a better overview.
Screenshot_2020-02-12 Select Podkategoria to change Django site admin
Screenshot_2020-02-12 Select user to change Django site admin
Screenshot_2020-02-12 Site administration Django site admin

**Comment by mimi89999** [comment]
I changed a bit UI colors:
Screenshot_2020-02-12 Site administration Django site admin

**Comment by mimi89999** [comment]
Please review. I think that most of the UI is themed now. I hope that there are no bugs left...

**Comment by mimi89999** [comment]
Was using Sass ever condidered? It would greatly simplify the code as variables could be defined and colors could be automatically calculated. It could also provide CSS minification.

**Comment by knyghty** [comment]
@mimi89999 Hi, I'm a big fan of this work 👍 

First, to answer your question, I think there's a desire to make contributions to Django easy, and while everyone who knows Sass also knows CSS, the reverse is not necessarily true. But I think the bigger reason is it'd introduce a build process that has to happen somewhere, either when developing, which people invariably forget about (this happened with compressing JS, so it is no longer a requirement since nobody was following it). The other option is to do it during `collectstatic` but then we'd need every user to have a Sass compiler installed, which I think would be out of the question.

On the other hand though, the admin only supports modern browsers now, so you _can_ use CSS variables and other modern CSS features. I realise they're a little clunkier but I have implemented dark themes with them fairly elegantly.

A couple of  asks from my side:

- The admin has seen some changes lately, particularly the new sidebar, but there have also been some other changes. It's probably worth rebasing and having a look through for anything that's now missing.
- Can you run the admin through an accessibility checker and make sure the contrast is good everywhere? I realise that the current default theme doesn't have good enough contrast in many places, and while I hope to remedy this soon, it'd be good to set off on the right foot here.

**Comment by knyghty** [comment]
@mimi89999, do you plan to continue this? I may take it up if not.

**Comment by mimi89999** [comment]
Hello,
Sorry for not replying. I was busy and then forgot about your comment.

> do you plan to continue this? I may take it up if not.

I think that it's almost finished. The core parts are themed. Some parts are still missing, but I will add them soon. There was no real feedback from project maintainers, so I didn't know if it's the way to go or I should change something.

> First, to answer your question, I think there's a desire to make contributions to Django easy, and while everyone who knows Sass also knows CSS, the reverse is not necessarily true.

I think that learning Sass doesn't take long. To make basic changes almost no knowledge of Sass is needed.

> But I think the bigger reason is it'd introduce a build process that has to happen somewhere, either when developing, which people invariably forget about (this happened with compressing JS, so it is no longer a requirement since nobody was following it). The other option is to do it during collectstatic but then we'd need every user to have a Sass compiler installed, which I think would be out of the question.

In a Django project I'm working on, we store compiled css in the repo and it's a nightmare because everybody forgets about updating it. Best would be to do it when submitting to pypi, but I don't know if it is possible.

> The admin has seen some changes lately, particularly the new sidebar, but there have also been some other changes. It's probably worth rebasing and having a look through for anything that's now missing.

I will update it.

> Can you run the admin through an accessibility checker and make sure the contrast is good everywhere? I realise that the current default theme doesn't have good enough contrast in many places, and while I hope to remedy this soon, it'd be good to set off on the right foot here.

I did run it through FF accessibility checker. I will check again after the update and will show you the results.

**Comment by knyghty** [comment]
> I didn't know if it's the way to go or I should change something.

It looks okay to me. Another way you could consider is to have a set of CSS variables, perhaps at the beginning of the main css file, or perhaps in its own file to make it easy to override.  and then override them, e.g.

```css
:root {
  --main-bg-color: #fff;
  ...
}

@media (prefers-color-scheme: dark) {
  --main-bg-color: #000;
  ...
}
```

This might end up being easier to maintain as you don't need to touch other files, but I think your approach is also fine.

**Comment by mimi89999** [comment]
I rebased this PR, updated it and added theming of the new sidebar.

**Comment by knyghty** [comment]
Hi, I've had a look through locally. It looks really good generally. I have found a few issues:

Firefox's accessibility check finds some contrast issues (with the button text, and all text in the filter box other than the title).
Screenshot 2020-08-19 at 22 30 13

---

This doesn't look beautiful. I know styling forms can be tricky, but can something be done here?
Screenshot 2020-08-19 at 22 31 11

---

I think there's a contrast issue with the label. I also think for a dark theme, this bright white background is quite jarring. Perhaps the buttons could be changed as well, but as they're SVGs it might not be worth it.
Screenshot 2020-08-19 at 22 32 02

---

These table headings need more contrast.
Screenshot 2020-08-19 at 22 33 10

---

Please also check admindocs. There are a few contrast issues and these code blocks should be made dark.
Screenshot 2020-08-19 at 22 33 49

---

Perhaps this could be a nicer colour, and the action "Go" button.
Screenshot 2020-08-19 at 22 35 29

---

I realise these buttons are disabled, but I think I should still be able to read them.
Screenshot 2020-08-19 at 22 47 36

---

I also think we should have the colours as CSS variables in one place as I think it'd be easier to maintain both themes side-by-side, but I'm not sure if it needs to be done here - @carltongibson do you perhaps have an opinion?

**Comment by mimi89999** [comment]
> Firefox's accessibility check finds some contrast issues (with the button text, and all text in the filter box other than the title).

I tested with webhint.io addon and it only reported issues with the header, which I changed. I will need to look deeper into this contrast issue.

> This doesn't look beautiful. I know styling forms can be tricky, but can something be done here?

On my DE it looks good. Is it even possible to set the background of that dropdown from CSS?
Capture d’écran du 2020-08-20 12-20-13

> I think there's a contrast issue with the label. I also think for a dark theme, this bright white background is quite jarring.

Widgets are not themed yet as reported previously. I will need to work on this.

I will fix other issues that those with widgets soon.

**Comment by knyghty** [comment]
> On my DE it looks good. Is it even possible to set the background of that dropdown from CSS?

It's browser-dependent. Firefox respects `background-color` and `color`. Chrome respects them but only when closed. For `background-color` it seems to choose a colour based on what you choose, and it chooses a `color` presumably to contrast with it. For example, with `background: blue; color: black;` on Chrome:

Screenshot 2020-08-20 at 12 45 31
Screenshot 2020-08-20 at 12 45 38

**Comment by matthiask** [comment]
I have missed this PR before starting possibly related work in https://github.com/django/django/pull/13435. This pull request proposes using CSS variables for colors. Implementing a dark mode should be a matter of setting a few CSS variables (see here https://github.com/django/django/commit/95b489f65dd0790fa1703ee38c8eee6d210a61ed#diff-9f1f1f2e0ee84be82ee1b3bd34f485a1L35-L69 )

**Comment by carltongibson** [comment]
Hi @mimi89999 — I intend to merge @matthiask's #13435 this morning. If you can rebase off that and have this ready before the feature freeze on 14th Jan, I would be **very happy** to pull this into Django 3.2. 😀

Thanks for your work!

**Comment by carltongibson** [comment]
Just unresolved the conversation about testing the fallback for Edge. (_Surely that's supported now it's Chromium based..._ 🤔)

We need to think about adding a note to the new Admin Theming (and the release note) mentioning that Dark Mode is available (and installed, assuming we do/can use the media query.)

**Comment by mimi89999** [comment]
Hello,
I started work on rebasing this PR, but encountered a blocking issue: https://github.com/django/django/pull/13435#discussion_r553245949

**Comment by mimi89999** [comment]
Replaced by https://github.com/django/django/pull/13880.

**Pull request closed without merging**